### PR TITLE
Implement patch whitelisting for zypper

### PR DESCRIPTION
**Pull request by iamsubratp**
--with-update is a tricky case to handle with patches. All the other fields/flags in zyppersettings work on patches except for --with-update, which runs on packages.
readmore: https://www.mankier.com/8/zypper#--with-update

Here is the algorithm to prepare patch and package list for zypper:

- Fetch the list of patches with the filters on severity, category
- If user specifies --with-update, fetch the list of packages that will be updated.
- Filter out any package that will be updated as a part of a patch install from the list of packages to be updated;
- to know packages are going to be updated as a part of a patch, we have used "zypper info -t " command
-  for the list of excluded patches specified by user, filter out the patches from the previously generated patch list.

### CODE REVIEW

File: ospatch/zypper_patch.go [hunk 1]
```diff
@@ -67,39 +73,124 @@ func ZypperUpdateWithUpdate(withUpdate bool) ZypperPatchOption {
...
+	for _, patch := range fPatches {
+		args = append(args, "patch:"+patch.Name)
 	}
-	for _, s := range zOpts.severities {
-		args = append(args, "--severity="+s)
+	for _, pkg := range fpkgs {
+		args = append(args, "package:"+pkg.Name)
 	}
 
 	if _, err := zOpts.runner(exec.Command(zypper, args...)); err != nil {
```
[the remark this answers is not in the document]
Comment (iamsubratp) [reply ospatch/zypper_patch.go]: makes sense. will change

File: ospatch/zypper_patch.go [hunk 2]
```diff
@@ -25,11 +28,14 @@ var (
...
-	withOptional bool
-	withUpdate   bool
-	runner       func(cmd *exec.Cmd) ([]byte, error)
+	categories       []string
+	severities       []string
+	withOptional     bool
+	withUpdate       bool
+	excludes         []string
+	exclusivePatches []string
+	runner           func(cmd *exec.Cmd) ([]byte, error)
```
Comment (adjackura) [inline ospatch/zypper_patch.go]: You can drop the runner now as we don't use it anymore

Comment (iamsubratp) [reply ospatch/zypper_patch.go]: done

File: ospatch/zypper_patch.go [hunk 3]
```diff
@@ -67,39 +71,100 @@ func ZypperUpdateWithUpdate(withUpdate bool) ZypperPatchOption {
...
-		args = append(args, "--with-update")
+	patches, err := packages.ZypperPatches(zListOpts...)
+	if err != nil {
+		return err
 	}
-	for _, c := range zOpts.categories {
-		args = append(args, "--category="+c)
+
+	fPatches, fpkgs, err := runFilter(patches, zOpts.exclusivePatches, zOpts.excludes, zOpts.withUpdate)
+	return packages.ZypperInstall(fPatches, fpkgs)
```
Comment (adjackura) [inline ospatch/zypper_patch.go]: Add logging and dryrun return like we have with the other systems:
https://github.com/GoogleCloudPlatform/osconfig/blob/master/ospatch/apt_upgrade.go#L89

Comment (iamsubratp) [reply ospatch/zypper_patch.go]: done

**Pull request merged**